### PR TITLE
RavenDB-19331: Implementation of key reuse scopes & other key encoding optimizations

### PR DESCRIPTION
**Pull request by redknightlois**
This is still a work in process. It is still not optimized, just looking to ensure it is correct.

### CODE REVIEW

**Comment by redknightlois** [comment]
Trying to figure out if I can get rid entirely of the hashing structure, that will provide much better memory locality and compact code.

File: src/Sparrow/Compression/VariableSizeEncoding.cs [hunk 1]
```diff
@@ -261,6 +261,51 @@ private static void ThrowNotEnoughOutputSpace()
...
+
+            int offset = 0;
+
+            if (typeof(T) == typeof(long) || typeof(T) == typeof(ulong) ||
+                typeof(T) == typeof(int) || typeof(T) == typeof(uint) ||
+                typeof(T) == typeof(short) || typeof(T) == typeof(ushort))
+            {
+                int bytesReadWhenOverflow;
+                if (typeof(T) == typeof(long) || typeof(T) == typeof(ulong))
+                    bytesReadWhenOverflow = 12;
```
Comment (ayende) [inline src/Sparrow/Compression/VariableSizeEncoding.cs]: variable len size of long is 9, for ulong, 10

File: src/Sparrow/Compression/VariableSizeEncoding.cs [hunk 2]
```diff
@@ -261,6 +261,51 @@ private static void ThrowNotEnoughOutputSpace()
...
+                else if (typeof(T) == typeof(int) || typeof(T) == typeof(uint))
+                    bytesReadWhenOverflow = 7;
```
Comment (ayende) [inline src/Sparrow/Compression/VariableSizeEncoding.cs]: For all of those, that should be 5, no?

File: src/Voron/Data/Sets/Set.cs [hunk 3]
```diff
@@ -630,10 +639,10 @@ public Iterator(Set parent)
 
                 // We need to find the long.MinValue therefore the fastest way is to always
                 // take the left-most pointer on any branch node.
-                _parent.FindSmallestValue();
+                var state = _parent.FindSmallestValue();
 
-                ref var state = ref _parent._stk[_parent._pos];
-                var leafPage = new SetLeafPage(state.Page);
+                //ref var state = ref _parent._stk[_parent._pos];
```
Comment (ayende) [inline src/Voron/Data/Sets/Set.cs]: can remove this?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 4]
```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
...
+                public int KeyIndex;
+            }
+
+            private int _currentKeyIdx;
+
+            private const int MappingTableSize = 64;
+            private const int MappingTableMask = MappingTableSize - 1;
+
+            private readonly long[] _keyMappingDictionary = new long[MappingTableSize];
+            private readonly int[] _keyMappingIndex = new int[MappingTableSize];
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: Those are used only _during_ the searching process, right?
Can we make the thread static, then? Would save some allocations.

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 5]
```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
...
+                {
+                    Debug.Assert(IsValid, "At this stage we either created the key using an unencoded version. Current dictionary cannot be invalid.");
+
+                    // We acquire the decoded form. This will lazily evaluate if needed. 
+                    ReadOnlySpan<byte> decodedKey = Decoded();
+
+                    // IMPORTANT: Pointers are potentially invalidated by the grow storage call but not the indexes.
+
+                    // We look for an appropriate place to put this key
+                    int buckedIdx = SelectHashBucketForWrite();
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: I though we aren't doing hashing any more?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 6]
```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
...
-                return Get(key, tree, dictionaryId);
+                return Invalid;
             }
 
-            public static EncodedKey Get(ReadOnlySpan<byte> key, CompactTree tree, long dictionaryId)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private int SelectHashBucketForWrite()
             {
-                var dictionary = tree.GetEncodingDictionary(dictionaryId);
+                _keyMappingCurrent++;
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: this is linear, so great, but we need a better name here, I think

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 5 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: Need to change the name, yes.

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 4 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: Yes. Haven't look yet at what allocation avoidance technique would be proper to use here.

**Comment by arekpalinski** [comment]
@redknightlois is it still draft or something that we could already merge?  (after we merge https://github.com/ravendb/ravendb/pull/15118 and confirm that everything works). Or you want to test it on real data before doing merge?

**Comment by redknightlois** [comment]
While still working on this, and therefore not ready for merging, would be
great if we can give it a run on a real-world. Working on optimizing the
key comparisons so when that is done and the results are I will ponder if
merging it as an intermediate step.

**Comment by arekpalinski** [comment]
Can you please fix the compilation and solve the conflict?

File: src/Sparrow/Compression/VariableSizeEncoding.cs (hunk 1 above)
Comment (redknightlois) [reply src/Sparrow/Compression/VariableSizeEncoding.cs]: No.

image

File: src/Sparrow/Compression/VariableSizeEncoding.cs (hunk 2 above)
Comment (redknightlois) [reply src/Sparrow/Compression/VariableSizeEncoding.cs]: They have to be the same as the other. They are all wrong by 2.

File: src/Corax/IndexWriter.cs [hunk 7]
```diff
@@ -1312,6 +1316,7 @@ private void InsertTextualField(Tree fieldsTree, IndexedField indexedField, Span
                         dumper.WriteRemoval(term, ttt);
                         break;
                 }
+                scope.Dispose();
```
Comment (ayende) [inline src/Corax/IndexWriter.cs:1324]: this looks weird
I assume you aren't using `using` here because of the `try/finaly` cost?

**Comment by arekpalinski** [comment]
Can you please check this test https://jenkins.hibernatingrhinos.com/job/PR_Tests/3773/testReport/junit/SlowTests.Corax.Bugs/CompactTreeAddAndRemove/gz__/ ?

File: src/Corax/IndexWriter.cs (hunk 7 above)
Comment (redknightlois) [reply src/Corax/IndexWriter.cs:1324]: Yes. We are avoiding the `try/finally` but returning it if possible.

**Pull request merged**